plone/app/vocabularies/catalog.py

```diff
@@ -646,6 +659,181 @@ def __call__(self, context, query=None):
+
+
+        directives.widget(
+            'my_relation',
+            AjaxSelectFieldWidget,
+            vocabulary=StaticCatalogVocabulary(
+                {'portal_type': 'Event'},
+                text_search_index="Title",
+                title_template="{brain.Type}: {brain.Title} at {path}"
+            )
+        )
+
+    This vocabulary lazily caches the result set for the base query on the
+    request to optimize performance.
+
+    Here are some doctests::
+
+      >>> from plone.app.vocabularies.tests.base import Brain
+      >>> from plone.app.vocabularies.tests.base import DummyCatalog
+      >>> from plone.app.vocabularies.tests.base import create_context
+      >>> from plone.app.vocabularies.tests.base import DummyTool
+
+      >>> context = create_context()
+
+      >>> catalog = DummyCatalog(('/1234', '/2345'))
+      >>> context.portal_catalog = catalog
+
+      >>> tool = DummyTool('portal_url')
+      >>> def getPortalPath():
+      ...     return '/'
+      >>> tool.getPortalPath = getPortalPath
+      >>> context.portal_url = tool
+
+      >>> vocab = StaticCatalogVocabulary({'portal_type': ['Document']})
+      >>> vocab
+      <plone.app.vocabularies.catalog.StaticCatalogVocabulary object at ...>
+
+      >>> vocab.search('')
+      <zope.schema.vocabulary.SimpleVocabulary object at ...>
+      >>> list(vocab.search(''))
+      []
+
+      >>> vocab.search('foo')
+      <zope.schema.vocabulary.SimpleVocabulary object at ...>
+
+      >>> [(t.title, t.value) for t in vocab.search('foo')]
+      [('BrainTitle (/1234)', '/1234'), ('BrainTitle (/2345)', '/2345')]
+
+    We strip out the site path from the rendered path in the title template:
+
+      >>> catalog = DummyCatalog(('/site/1234', '/site/2345'))
+      >>> context.portal_catalog = catalog
+      >>> vocab = StaticCatalogVocabulary({'portal_type': ['Document']})
+      >>> [(t.title, t.value) for t in vocab.search('bar')]
+      [('BrainTitle (/site/1234)', '/site/1234'),
+       ('BrainTitle (/site/2345)', '/site/2345')]
+
+      >>> context.__name__ = 'site'
+      >>> vocab = StaticCatalogVocabulary({'portal_type': ['Document']})
+      >>> [(t.title, t.value) for t in vocab.search('bar')]
+      [('BrainTitle (/1234)', '/site/1234'),
+       ('BrainTitle (/2345)', '/site/2345')]
+
+    The title template can be customized:
+
+      >>> vocab.title_template = "{url} {brain.UID} - {brain.Title} {path}"
+      >>> [(t.title, t.value) for t in vocab.search('bar')]
+      [('proto:/site/1234 /site/1234 - BrainTitle /1234', '/site/1234'),
+       ('proto:/site/2345 /site/2345 - BrainTitle /2345', '/site/2345')]
+
+    """
+    title_template = "{brain.Title} ({path})"
+    text_search_index = "SearchableText"
+
+    def __init__(self, query, text_search_index=None,
+                 title_template=None):
+        self.query = query
+        if text_search_index:
+            self.text_search_index = text_search_index
```

Comment:
This is a strange construct.
I wonder if:
```python
     def __init__(self, query, text_search_index="SearchableText"):
        self.query = query
        self.text_search_index = text_search_index
```
would be better.

About the `title_template` customization, I am not really sure it is super useful.

The title template depends highly on the format parameters.
IMO if you want to provide a flexible way to do that use a `brain2title` method that everybody can customize at will.

I know that this is more complex than just passing a parameter to the constructor, but I have the feeling this is going to cause less headaches in the future.

Anyway this is just my personal opinion, not a blocker

Reply:
I put the `None` defaults in the constructor so that subclasses didn't need to override `__init__` to change the defaults, but it's probably an unnecessary complication, since the class is intended to provide significant flexibility without subclassing.

I think there's quite a bit of power in being able to simply change `title_template`. In the simplest case it would just be `{brain.Title}` when there is a very limited set of values in the vocabulary and there is no need to disambiguate by path. I would say full access to all attributes of `brain` in a format string gets you quite a bit of power, adding in the full url and relative path gets most of what might be needed for even an HTML rendering of the title.

To me, the goal of this class is to provide significant flexibility without needing to subclass, but some cases may still need a subclass. More complex use cases could always sub-class and override `createTerm`.